extensions/typescript/src/features/jsDocCompletionProvider.ts

```diff
@@ -0,0 +1,184 @@
+		const suffix = line.slice(position.character).match(/^\s*\**\/?$/);
+		return suffix !== null;
+	}
+
+	/**
+	 * Try to insert a jsdoc comment, using a template provide by typescript
+	 * if possible, otherwise falling back to a default comment format.
+	 */
+	private tryCompleteJsDoc(resource: Uri, position: Position): Thenable<boolean> {
+		const file = this.client.normalizePath(resource);
+		if (!file) {
+			return Promise.resolve(false);
+		}
+
+		const editor = window.activeTextEditor;
+		if (!editor || editor.document.uri.fsPath !== resource.fsPath) {
+			return Promise.resolve(false);
+		}
+
+		return this.prepForDocCompletion(editor, position)
+			.then((start: Position) => {
+				return this.tryInsertJsDocFromTemplate(editor, file, start);
+			})
+			.then((didInsertFromTemplate: boolean) => {
+				if (didInsertFromTemplate) {
+					return true;
+				}
+				return this.tryInsertDefaultDoc(editor, position);
+			});
+	}
+
+	/**
+	 * Prepare the area around the position for insertion of the jsdoc.
+	 *
+	 * Removes any the prefix and suffix of a possible jsdoc
+	 */
+	private prepForDocCompletion(editor: TextEditor, position: Position): Thenable<Position> {
+		const line = editor.document.lineAt(position.line).text;
+		const prefix = line.slice(0, position.character).match(/\/\**\s*$/);
+		const suffix = line.slice(position.character).match(/^\s*\**\//);
+		if (!prefix && !suffix) {
+			// Nothing to remove
+			return Promise.resolve(position);
+		}
+
+		const start = position.translate(0, prefix ? -prefix[0].length : 0);
+		return editor.edit(
+			edits => {
+				edits.delete(new Range(start, position.translate(0, suffix ? suffix[0].length : 0)));
+			}, {
+				undoStopBefore: true,
+				undoStopAfter: false
+			}).then(() => start);
+	}
+
+	private tryInsertJsDocFromTemplate(editor: TextEditor, file: string, position: Position): Promise<boolean> {
+		const args: FileLocationRequestArgs = {
+			file: file,
+			line: position.line + 1,
+			offset: position.character + 1
+		};
+		return this.client.execute('docCommentTemplate', args)
+			.then((res: DocCommandTemplateResponse) => {
+				if (!res || !res.body) {
+					return false;
+				}
+				const commentText = res.body.newText;
+				return editor.edit(
+					edits => edits.insert(position, commentText),
+					{ undoStopBefore: false, undoStopAfter: true });
+			}, () => false)
+			.then((didInsertComment: boolean) => {
+				if (didInsertComment) {
+					const newCursorPosition = new Position(position.line + 1, editor.document.lineAt(position.line + 1).text.length);
+					editor.selection = new Selection(newCursorPosition, newCursorPosition);
+				}
+				return didInsertComment;
+			});
+	}
+
+	/**
+	 * Insert the default JSDoc
+	 */
+	private tryInsertDefaultDoc(editor: TextEditor, position: Position): Thenable<boolean> {
+		const line = editor.document.lineAt(position.line).text;
+		const spaceBefore = line.slice(0, position.character).match(/^\s*$/);
```

Comment:
When you use [`SnippetString`](https://github.com/Microsoft/vscode/blob/master/src/vs/vscode.d.ts#L2122) you get indentation for free and you can also define placeholders for params and define the final cursor position.